### ISSUE TITLE
blank results when rendering with large batch size

### ISSUE THREAD

**Issue by l1346792580123** [NVlabs/nvdiffrast#210]
I have encountered an problem when rendering with large batch size. The rendering code as follows

```
import pickle
import numpy as np
import cv2
from sklearn.preprocessing import normalize
import torch
import nvdiffrast.torch as dr

def lookat2extrinsic(origin, target, up):
    z = normalize((target-origin)[None]).reshape(-1)
    x = normalize(np.cross(up, z)[None]).reshape(-1)
    y = normalize(np.cross(z,x)[None]).reshape(-1)
    minv = np.eye(4)
    tr = np.eye(4)
    minv[0:1,:3] = x
    minv[1:2,:3] = y
    minv[2:3,:3] = z
    tr[:3,3] = -origin

    check = np.eye(4)
    check[:3,0] = x
    check[:3,1] = y
    check[:3,2] = z
    check[:3,3] = origin

    return np.matmul(minv, tr)

def build_cams2(num = 8):
    w2cs = []
    projs = []

    # body camera
    target = np.array([0,-0.4,0])
    up = np.array([0,-1,0])
    proj = np.zeros((4,4),dtype=np.float32)
    proj[0,0] = 2.
    proj[0,2] = 0.
    proj[1,1] = 2.
    proj[1,2] = 0.
    proj[2,2] = 0.
    proj[2,3] = -0.1
    proj[3,2] = 1.
    proj[3,3] = 0.
    for i in range(num):
        origin = np.array([2.2*np.sin(i/num*2*np.pi),-0.3,2.2*np.cos(i/num*2*np.pi)])
        w2c = lookat2extrinsic(origin,target,up).astype(np.float32)

        w2cs.append(w2c)
        projs.append(proj)

    # face camera
    target = np.array([0,0.1,0])
    up = np.array([0,-1,0])
    proj = np.zeros((4,4),dtype=np.float32)
    proj[0,0] = 2.
    proj[0,2] = 0.
    proj[1,1] = 2.
    proj[1,2] = 0.
    proj[2,2] = 0.
    proj[2,3] = -0.1
    proj[3,2] = 1.
    proj[3,3] = 0.

    for i in range(num):
        origin = np.array([1.2*np.sin(i/num*2*np.pi),0.1,1.2*np.cos(i/num*2*np.pi)])
        w2c = lookat2extrinsic(origin,target,up).astype(np.float32)

        w2cs.append(w2c)
        projs.append(proj)

    # lower camera
    target = np.array([0,-0.7,0])
    up = np.array([0,-1,0])
    proj = np.zeros((4,4),dtype=np.float32)
    proj[0,0] = 2.
    proj[0,2] = 0.
    proj[1,1] = 2.
    proj[1,2] = 0.
    proj[2,2] = 0.
    proj[2,3] = -0.1
    proj[3,2] = 1.
    proj[3,3] = 0.

    for i in range(num):
        origin = np.array([1.5*np.sin(i/num*2*np.pi),-0.7,1.5*np.cos(i/num*2*np.pi)])
        w2c = lookat2extrinsic(origin,target,up).astype(np.float32)

        w2cs.append(w2c)
        projs.append(proj)

    w2cs = torch.from_numpy(np.stack(w2cs)).permute(0,2,1).cuda()
    projs = torch.from_numpy(np.stack(projs)).permute(0,2,1).cuda()

    return w2cs, projs


def read_obj(obj_path):

    vertices = []
    faces = []
    uvs = []
    uv_faces = []
    v_inds = dict()
    with open(obj_path, 'r') as f:
        for line in f:
            if line.startswith('v '):
                verts = line.strip().split(' ')[1:]
                vertices.append(np.array([float(item) for item in verts]))
            elif line.startswith('f '):
                ff = line.strip().split(' ')[1:]
                face = []
                for item in ff:
                    f1, f2 = item.split('/')[:2]
                    f1 = int(f1) - 1
                    f2 = int(f2) - 1
                    if f2 in v_inds.keys():
                        assert f1 == v_inds[f2]
                    else:
                        v_inds[f2] = f1
                    # assert int(f1) == int(f2)
                    face.append(f2)
                faces.append(np.array(face).astype(int))

            elif line.startswith('vt '):
                vt = line.strip().split(' ')[1:]
                uvs.append(np.array([float(item) for item in vt]))

    num_verts = len(v_inds)
    vertices = [vertices[v_inds[i]] for i in range(num_verts)]
    vertices = np.stack(vertices)
    faces = np.stack(faces)
    uvs = np.stack(uvs)

    return vertices, faces, uvs


if __name__ == '__main__':

    h,w = 512,512
    batch = 24
    resolution = (h,w)


    glctx = dr.RasterizeGLContext()

    w2cs, projs = build_cams2()
    cam_num = w2cs.shape[0]

    smplx_param = pickle.load(open('smplx_param.pkl','rb'), encoding='latain1')

    vertices, faces, uv = read_obj('0000.obj')

    tex_img = cv2.imread('material0.jpeg')
    tex = torch.from_numpy((tex_img[::-1,:,:]/255.).astype(np.float32)).cuda()

    vertices /= smplx_param['scale']
    vertices -= smplx_param['transl']

    vertices = torch.from_numpy(vertices.astype(np.float32)).cuda()
    faces = torch.from_numpy(faces.astype(np.int32)).cuda()
    uv = torch.from_numpy(uv.astype(np.float32)).cuda()
    uv_faces = faces

    for idx in range(0,cam_num, batch):
        num = min(cam_num, idx+batch) - idx
        vertsw = torch.cat([vertices, torch.ones_like(vertices[:,0:1])], axis=1).unsqueeze(0).expand(num,-1,-1)
        rot_verts = torch.einsum('ijk,ikl->ijl', vertsw, w2cs[idx:idx+batch])
        proj_verts = torch.einsum('ijk,ikl->ijl', rot_verts, projs[idx:idx+batch])

        rast_out, rast_out_db = dr.rasterize(glctx, proj_verts, faces, resolution=resolution)

        texc, texd = dr.interpolate(uv[None].expand(num,-1,-1).contiguous(), rast_out, uv_faces)
        rast_albedo = dr.texture(tex[None].expand(num,-1,-1,-1).contiguous(), texc, filter_mode='linear')

        valid_idx = (rast_out[:,:,:,3] > 0)
        rast_albedo[~valid_idx]=0
        rast_albedo = (rast_albedo.squeeze().detach().cpu().numpy()*255).astype(np.uint8)
        masks = (valid_idx.squeeze().detach().cpu().numpy()*255).astype(np.uint8)

        for j in range(num):
            cv2.imwrite('%02d.png'%(idx+j), rast_albedo[j])
```

When I render with a batch size of 8, all rendering results are normal. While when I use a batch size of 24 for rendering, the rendering results of the 12th to 24th are problematic. The problematic rendering results are shown as follows

![Image](https://github.com/user-attachments/assets/9c08fe8f-f04e-45a7-ac00-a7bc9c12dfca)

![Image](https://github.com/user-attachments/assets/caad9809-accd-4f95-befd-e48c508517d5)

Is this a bug?  the rendering code and mesh can be found [here](https://github.com/user-attachments/files/18471904/test.zip).

**Comment by s-laine** [comment]
Yes, this is a bug caused by an overflow in texture indexing. Your example code multiplies the texture size by batch size, creating a tensor large enough to overflow an internal 32-bit index in the `dr.texture()` call.

As a workaround, if you don't need the texture to be unique for each minibatch element, you can replace `tex[None].expand(num,-1,-1,-1).contiguous()` by `tex[None]` and rely on broadcasting along minibatch axis. This keeps the texture tensor size well below the limit. Otherwise, you'll have to run the texture op in batches that keep the total number of elements in the texture below 2^31.

A proper fix would require modifications to the texture op internals. I'll add that in the TODO list.

**Comment by l1346792580123** [comment]
Thanks for for your reply !
